tools/SourceKit/lib/SwiftLang/SwiftTypeContextInfo.cpp

```diff
@@ -0,0 +1,246 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "SwiftASTManager.h"
+#include "SwiftLangSupport.h"
+#include "swift/Frontend/Frontend.h"
+#include "swift/Frontend/PrintingDiagnosticConsumer.h"
+#include "swift/IDE/TypeContextInfo.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Comment.h"
+#include "clang/AST/Decl.h"
+
+using namespace SourceKit;
+using namespace swift;
+using namespace ide;
+
+/// Copy a memory buffer inserting '0' at the position of \c origBuf.
+// TODO: Share with code completion.
+static std::unique_ptr<llvm::MemoryBuffer>
+makeCodeCompletionMemoryBuffer(const llvm::MemoryBuffer *origBuf,
+                               unsigned &Offset,
+                               const std::string bufferIdentifier) {
+
+  auto origBuffSize = origBuf->getBufferSize();
+  if (Offset > origBuffSize)
+    Offset = origBuffSize;
+
+  auto newBuffer = llvm::WritableMemoryBuffer::getNewUninitMemBuffer(
+      origBuffSize + 1, bufferIdentifier);
+  auto *pos = origBuf->getBufferStart() + Offset;
+  auto *newPos =
+      std::copy(origBuf->getBufferStart(), pos, newBuffer->getBufferStart());
+  *newPos = '\0';
+  std::copy(pos, origBuf->getBufferEnd(), newPos + 1);
+
+  return std::unique_ptr<llvm::MemoryBuffer>(newBuffer.release());
+}
+
+static bool swiftTypeContextInfoImpl(SwiftLangSupport &Lang,
+                                     llvm::MemoryBuffer *UnresolvedInputFile,
+                                     unsigned Offset,
+                                     ArrayRef<const char *> Args,
+                                     ide::TypeContextInfoConsumer &Consumer,
+                                     std::string &Error) {
+  auto bufferIdentifier =
+      Lang.resolvePathSymlinks(UnresolvedInputFile->getBufferIdentifier());
+
+  auto newBuffer = makeCodeCompletionMemoryBuffer(UnresolvedInputFile, Offset,
+                                                  bufferIdentifier);
+
+  CompilerInstance CI;
+  PrintingDiagnosticConsumer PrintDiags;
+  CI.addDiagnosticConsumer(&PrintDiags);
+
+  CompilerInvocation Invocation;
+  bool Failed = Lang.getASTManager()->initCompilerInvocation(
+      Invocation, Args, CI.getDiags(), bufferIdentifier, Error);
+  if (Failed)
+    return false;
+  if (!Invocation.getFrontendOptions().InputsAndOutputs.hasInputs()) {
+    Error = "no input filenames specified";
+    return false;
+  }
+
+  Invocation.setCodeCompletionPoint(newBuffer.get(), Offset);
+
+  // Create a factory for code completion callbacks that will feed the
+  // Consumer.
+  std::unique_ptr<CodeCompletionCallbacksFactory> callbacksFactory(
+      ide::makeTypeContextInfoCallbacksFactory(Consumer));
+
+  Invocation.setCodeCompletionFactory(callbacksFactory.get());
+
+  if (CI.setup(Invocation)) {
+    // FIXME: error?
+    return true;
+  }
+  CI.performSema();
+
+  return true;
```

Comment:
We should copy over the `trace::TracedOperation` code from code-completion as well so that this action shows up in compilation traces such as the Index Log.